### PR TITLE
Add map Generator algorithm

### PR DESCRIPTION
**Pull request by Yinonss**
Change map generator algorithm
(4b scene)

### CODE REVIEW

File: Assets/Scripts/4-generation/CaveGenerator.cs [hunk 1]
```diff
@@ -56,9 +58,10 @@ public class CaveGenerator {
                 if (x == 0 || x == gridSize - 1 || y == 0 || y == gridSize - 1) {
                     //We dont want holes in our walls, so the border is always a wall
                     bufferOld[x, y] = 1;
-                } else {
+                }                
+                else {
                     //Random walls and caves
-                    bufferOld[x, y] = random.NextDouble() < randomFillPercent ? 1 : 0;
+                    bufferOld[x, y] = Random.Range(0, 4);
```
Comment (DYMGames) [inline Assets/Scripts/4-generation/CaveGenerator.cs]: configure the number 4 as readonly variable

File: Assets/Scripts/4-generation/CaveGenerator.cs [hunk 2]
```diff
@@ -77,44 +80,46 @@ public class CaveGenerator {
...
-                //Use some smoothing rules to generate caves
-                if (surroundingWalls > 4) {
-                    bufferNew[x, y] = 1;
-                } else if (surroundingWalls == 4) {
-                    bufferNew[x, y] = bufferOld[x, y];
-                } else {
-                    bufferNew[x, y] = 0;
-                }
+                bufferNew[x, y] = SurroundingCheck(x, y);
+                Console.WriteLine(bufferNew[x, y]);
```
Comment (DYMGames) [inline Assets/Scripts/4-generation/CaveGenerator.cs]: Delete

File: Assets/Scripts/4-generation/CaveGenerator.cs [hunk 3]
```diff
@@ -77,44 +80,46 @@ public class CaveGenerator {
...
-        for (int neighborX = cellX - 1; neighborX <= cellX + 1; neighborX ++) {
-            for (int neighborY = cellY - 1; neighborY <= cellY + 1; neighborY++) {
-                //We dont need to care about being outside of the grid because we are never looking at the border
-                //This is the cell itself and no neighbor!
-                if (neighborX == cellX && neighborY == cellY) {
+    // Check which tile is more common around cell [x, y] and change it to match the majority if needed.
+    private int SurroundingCheck(int cellX, int cellY)
+    {
+        int[] countTiles = { 0, 0, 0, 0};
+        for (int i = cellX - 1; i < cellX + 2; i++)
```
Comment (DYMGames) [inline Assets/Scripts/4-generation/CaveGenerator.cs:96]: Why + 2 ? why + 1?

File: Assets/Scripts/4-generation/CaveGenerator.cs [hunk 4]
```diff
@@ -77,44 +80,46 @@ public class CaveGenerator {
...
+        {
+            for (int j = cellY - 1; j < cellY + 2; j++)
```
Comment (DYMGames) [inline Assets/Scripts/4-generation/CaveGenerator.cs:98]: why +2 and -1 ?

File: Assets/Scripts/4-generation/CaveGenerator.cs [hunk 5]
```diff
@@ -77,44 +80,46 @@ public class CaveGenerator {
...
                     continue;
-                }
-
-                //This neighbor is a wall
-                if (bufferOld[neighborX, neighborY] == 1) {
-                    wallCounter += 1;
+                else
+                {
+                    switch(bufferOld[i, j])
+                    {
```
Comment (DYMGames) [inline Assets/Scripts/4-generation/CaveGenerator.cs]: bad coding.
Just replace with
 countTiles[bufferOld[i, j]]++;

File: Assets/Scripts/4-generation/CaveGenerator.cs [hunk 6]
```diff
@@ -77,44 +80,46 @@ public class CaveGenerator {
...
+                        case 1: countTiles[1]++; break;
+                        case 2: countTiles[2]++; break;
+                        case 3: countTiles[3]++; break;
+                    }
                 }
             }
         }
-        return wallCounter;
+        // If the tile has at least three similar tiles around - do not change it.
+        if (countTiles[bufferOld[cellX, cellY]] > 3)
```
Comment (DYMGames) [inline Assets/Scripts/4-generation/CaveGenerator.cs]: why 3 ?

Comment (DYMGames) [reply Assets/Scripts/4-generation/CaveGenerator.cs]: use as readonly number

File: Assets/Scripts/4-generation/TilemapCaveGenerator.cs [hunk 7]
```diff
@@ -19,6 +19,12 @@ public class TilemapCaveGenerator: MonoBehaviour {
     [Tooltip("The tile that represents a floor (a passable block)")]
     [SerializeField] TileBase floorTile = null;
 
+    [Tooltip("The tile that represents a wall (an impassable block)")]
+    [SerializeField] TileBase waterTile = null;
```
Comment (DYMGames) [inline Assets/Scripts/4-generation/TilemapCaveGenerator.cs]: Please store and manage them in array

**Review by DYMGames — CHANGES_REQUESTED** [review]

File: Assets/Scripts/4-generation/CaveGenerator.cs (hunk 5 above)
Comment (Yinonss) [reply Assets/Scripts/4-generation/CaveGenerator.cs]: nice

File: Assets/Scripts/4-generation/TilemapCaveGenerator.cs [hunk 8]
```diff
@@ -65,13 +71,20 @@ public class TilemapCaveGenerator: MonoBehaviour {
...
-    //Generate a black or white texture depending on if the pixel is cave or wall
+    //Generate different texture depending on if the pixel is floor, wall , water or forest.
     //Display the texture on a plane
     private void GenerateAndDisplayTexture(int[,] data) {
         for (int y = 0; y < gridSize; y++) {
             for (int x = 0; x < gridSize; x++) {
                 var position = new Vector3Int(x, y, 0);
-                var tile = data[x, y] == 1 ? wallTile: floorTile;
+                var tile = floorTile;              
+                switch (data[x, y])
```
Comment (DYMGames) [inline Assets/Scripts/4-generation/TilemapCaveGenerator.cs]: replace with
tile = arrayOfTiles[data[x, y]]

**Review by DYMGames — CHANGES_REQUESTED** [review]

File: Assets/Scripts/4-generation/CaveGenerator.cs (hunk 4 above)
Comment (Yinonss) [reply Assets/Scripts/4-generation/CaveGenerator.cs:98]: Essential , apparently.

**Review by DYMGames — APPROVED** [review]

**Pull request merged**